File: examples/StereoVOExample_large.cpp
/* ----------------------------------------------------------------------------

* GTSAM Copyright 2010, Georgia Tech Research Corporation,
* Atlanta, Georgia 30332-0415
* All Rights Reserved
* Authors: Frank Dellaert, et al. (see THANKS for the full author list)

* See LICENSE for the license information

* -------------------------------------------------------------------------- */

/**
* @file StereoVOExample_large.cpp
* @brief A stereo visual odometry example
* @date May 25, 2014
* @author Stephen Camp
*/


/**
 * A 3D stereo visual odometry example
 *  - robot starts at origin
 *  -moves forward, taking periodic stereo measurements
 *  -takes stereo readings of many landmarks
 */

#include <gtsam/geometry/Pose3.h>
#include <gtsam/geometry/Cal3_S2Stereo.h>
#include <gtsam/nonlinear/Values.h>
#include <gtsam/nonlinear/NonlinearEquality.h>
#include <gtsam/nonlinear/NonlinearFactorGraph.h>
#include <gtsam/nonlinear/LevenbergMarquardtOptimizer.h>
#include <gtsam/inference/Symbol.h>
#include <gtsam/slam/StereoFactor.h>
#include <gtsam/slam/dataset.h>

#include <string>
#include <fstream>
#include <iostream>

using namespace std;
using namespace gtsam;

int main(int argc, char** argv) {
  Values initial_estimate;
  NonlinearFactorGraph graph;
  const auto model = noiseModel::Isotropic::Sigma(3, 1);

  string calibration_loc = findExampleDataFile("VO_calibration.txt");
  string pose_loc = findExampleDataFile("VO_camera_poses_large.txt");
  string factor_loc = findExampleDataFile("VO_stereo_factors_large.txt");

  // read camera calibration info from file
  // focal lengths fx, fy, skew s, principal point u0, v0, baseline b
  double fx, fy, s, u0, v0, b;
  ifstream calibration_file(calibration_loc.c_str());
  cout << "Reading calibration info" << endl;
  calibration_file >> fx >> fy >> s >> u0 >> v0 >> b;

  // create stereo camera calibration object
  const Cal3_S2Stereo::shared_ptr K(new Cal3_S2Stereo(fx, fy, s, u0, v0, b));

  ifstream pose_file(pose_loc.c_str());
  cout << "Reading camera poses" << endl;
  int pose_id;
  MatrixRowMajor m(4, 4);
  // read camera pose parameters and use to make initial estimates of camera
  // poses
  while (pose_file >> pose_id) {
    for (int i = 0; i < 16; i++) {
      pose_file >> m.data()[i];
    }
    initial_estimate.insert(Symbol('x', pose_id), Pose3(m));
  }

  // camera and landmark keys
  size_t x, l;

  // pixel coordinates uL, uR, v (same for left/right images due to
  // rectification) landmark coordinates X, Y, Z in camera frame, resulting from
  // triangulation
  double uL, uR, v, X, Y, Z;
  ifstream factor_file(factor_loc.c_str());
  cout << "Reading stereo factors" << endl;
  // read stereo measurement details from file and use to create and add
  // GenericStereoFactor objects to the graph representation
  while (factor_file >> x >> l >> uL >> uR >> v >> X >> Y >> Z) {
    graph.emplace_shared<GenericStereoFactor<Pose3, Point3> >(
        StereoPoint2(uL, uR, v), model, Symbol('x', x), Symbol('l', l), K);
    // if the landmark variable included in this factor has not yet been added
    // to the initial variable value estimate, add it
    if (!initial_estimate.exists(Symbol('l', l))) {
      Pose3 camPose = initial_estimate.at<Pose3>(Symbol('x', x));
      // transformFrom() transforms the input Point3 from the camera pose space,
      // camPose, to the global space
      Point3 worldPoint = camPose.transformFrom(Point3(X, Y, Z));
      initial_estimate.insert(Symbol('l', l), worldPoint);
    }
  }

  Pose3 first_pose = initial_estimate.at<Pose3>(Symbol('x', 1));
  // constrain the first pose such that it cannot change from its original value
  // during optimization
  // NOTE: NonlinearEquality forces the optimizer to use QR rather than Cholesky
  // QR is much slower than Cholesky, but numerically more stable
  graph.emplace_shared<NonlinearEquality<Pose3> >(Symbol('x', 1), first_pose);

  cout << "Optimizing" << endl;
  // create Levenberg-Marquardt optimizer to optimize the factor graph
  LevenbergMarquardtParams params;
  params.orderingType = Ordering::METIS;
  LevenbergMarquardtOptimizer optimizer(graph, initial_estimate, params);
  Values result = optimizer.optimize();

  cout << "Final result sample:" << endl;
  Values pose_values = result.filter<Pose3>();
  pose_values.print("Final camera poses:\n");

  return 0;
}

File: matlab/gtsam_examples/StereoVOExample_large.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% GTSAM Copyright 2010, Georgia Tech Research Corporation,
% Atlanta, Georgia 30332-0415
% All Rights Reserved
% Authors: Frank Dellaert, et al. (see THANKS for the full author list)
%
% See LICENSE for the license information
%
% @brief Read Stereo Visual Odometry from file and optimize
% @author Chris Beall
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

import gtsam.*

%% Load calibration
% format: fx fy skew cx cy baseline
calib = dlmread(findExampleDataFile('VO_calibration.txt'));
K = Cal3_S2Stereo(calib(1), calib(2), calib(3), calib(4), calib(5), calib(6));
stereo_model = noiseModel.Diagonal.Sigmas([1.0; 1.0; 1.0]);

%% create empty graph and values
graph = NonlinearFactorGraph;
initial = Values;


%% load the initial poses from VO
% row format: camera_id 4x4 pose (row, major)
fprintf(1,'Reading data\n');
cameras = dlmread(findExampleDataFile('VO_camera_poses_large.txt'));
for i=1:size(cameras,1)
    pose = Pose3(reshape(cameras(i,2:17),4,4)');
    initial.insert(symbol('x',cameras(i,1)),pose);
end

%% load stereo measurements and initialize landmarks
% camera_id landmark_id uL uR v X Y Z
measurements = dlmread(findExampleDataFile('VO_stereo_factors_large.txt'));

fprintf(1,'Creating Graph\n'); tic
for i=1:size(measurements,1)
    sf = measurements(i,:);
    graph.add(GenericStereoFactor3D(StereoPoint2(sf(3),sf(4),sf(5)), stereo_model, ...
        symbol('x', sf(1)), symbol('l', sf(2)), K));
    
    if ~initial.exists(symbol('l',sf(2)))
        % 3D landmarks are stored in camera coordinates: transform
        % to world coordinates using the respective initial pose
        pose = initial.atPose3(symbol('x', sf(1)));
        world_point = pose.transformFrom(Point3(sf(6),sf(7),sf(8)));
        initial.insert(symbol('l',sf(2)), world_point);
    end
end
toc

%% add a constraint on the starting pose
key = symbol('x',1);
first_pose = initial.atPose3(key);
graph.add(NonlinearEqualityPose3(key, first_pose));

%% optimize
fprintf(1,'Optimizing\n'); tic
optimizer = LevenbergMarquardtOptimizer(graph, initial);
result = optimizer.optimizeSafely();
toc

%% visualize initial trajectory, final trajectory, and final points
cla; hold on;

plot3DTrajectory(initial, 'r', 1, 0.5);
plot3DTrajectory(result, 'g', 1, 0.5);
plot3DPoints(result);

axis([-5 20 -20 20 0 100]);
axis equal
view(3)
camup([0;1;0]);
